### DISCUSSION TITLE
pgbouncer in a netbox-branching environment

### DISCUSSION THREAD

**Discussion by tyler-8** [netboxlabs/netbox-branching#191] [General]
I use the connection-pooler `pgbouncer` in my NetBox environment and went to start using `netbox-branching` and have run into issues so wanted to share my findings/notes here in hopes that a solution can be found.

Upon launching NetBox with the plugin installed, this error is raised:

`unsupported startup parameter in options: search_path`

That had me searching around and I found this relevant issue/comment: https://github.com/netboxlabs/netbox-branching/issues/165#issuecomment-2454667365

and some other related discussions around the internet:

- https://stackoverflow.com/a/53890386
- https://github.com/pgbouncer/pgbouncer/issues/89
- https://github.com/pgbouncer/pgbouncer/pull/73

Some of the "solutions" out there say you can add one of these options to your pgbouncer config:

```
track_extra_parameters = search_path
ignore_startup_parameters = search_path
```

and that does get things partially working. NetBox loads, and branches can be created. Creating a new branch in NetBox shows "`Changes Ahead`" counts/lists populated with data - but it's a brand new branch off of main so there shouldn't be _any_ changes to speak of.

Removing pgbouncer from the equation and having NetBox connect directly to the database makes this issue go away. So clearly without `search_path` support today, netbox-branching appears to not do a real diff.

I attempted merging the branch anyway, just to see what happens, and it failed due to the `Device matching query does not exist.` - since in reality, that device had already been deleted in the past in main.

Abbreviated log output of attempted merge:

```{
    "log": [
        "Merging branch Tinkering (branch_jhczvebb)",
        "Found 897 changes to merge",
        "Setting branch status to merging",
        "Applying change DCIM | interface TenGigabitEthernet1/1/3 deleted by netUser using default",
        "interface ID 332560 already deleted; skipping",
        "Applying change DCIM | interface TenGigabitEthernet1/1/7 deleted by netUser using default",
        "interface ID 332593 already deleted; skipping",
        "Applying change DCIM | interface FastEthernet1 deleted by netUser using default",
        "interface ID 332595 already deleted; skipping",
        "Applying change DCIM | interface TenGigabitEthernet1/1/14 deleted by netUser using default",
        "interface ID 332606 already deleted; skipping",
        "Applying change DCIM | device rtr1.office-a updated by netUser using default",
        "Device matching query does not exist."
    ]
}
```

**Comment by tyler-8** [comment]
There might be some options available or coming soon.

- Supposedly the `citus` extension for PostgreSQL supports `track_extra_parameters = search_path` today. 
- A mention of some future patches to PostgreSQL itself that would allow the above behavior natively https://github.com/pgbouncer/pgbouncer/pull/867#issuecomment-1975186502 and https://github.com/pgbouncer/pg_pgbouncer/issues/5 and https://www.postgresql.org/message-id/flat/CAFh8B%3Dk8s7WrcqhafmYhdN1%2BE5LVzZi_QaYDq8bKvrGJTAhY2Q%40mail.gmail.com

**Reply by tyler-8** [reply]
A small update - so it does appear the 'fix' for allowing `track_extra_parameters = search_path` has been merged into the PostgreSQL codebase, but it has not yet been merged to any currently released versions as far as I can tell (17.2 is the latest as of this writing).

**Reply by tyler-8** [reply]
It looks like the change is in Postgres 18 (beta), slated for release in September 2025.

https://www.postgresql.org/docs/18/release-18.html#RELEASE-18-LIBPQ
https://www.postgresql.org/developer/roadmap/

**Reply by blipnet** [reply]
Did Postgres 18 solve the problems with pgbouncer?
I'm still struggling with getting branches to work as expected

**Reply by tyler-8** [reply]
I haven't had the opportunity to test it yet. I'm fairly certain you'd still need to add `track_extra_parameters = search_path` to your pgbouncer configuration, the difference, in theory, with Postgres 18 is that it will actually listen to that and provide the search path in responses.

**Comment by tyler-8** [comment]
I managed to get some (brief) testing with Postgres 18 and pgbouncer 1.25 configured with `track_extra_parameters = search_path` and branching appears to be working as expected. I encourage others to do their own and more rigorous testing, but so far I'm not seeing the previous behavior I documented above.

Tested with:

- `netbox-branching 0.7.2`
- `PostgreSQL 18.0`
- `pgbouncer 1.25.0`
- `NetBox v4.4.5`
